providers/multiprovider/src/main/java/dev/openfeature/contrib/providers/multiprovider/FirstMatchStrategy.java

```diff
@@ -0,0 +1,50 @@
+package dev.openfeature.contrib.providers.multiprovider;
+
+import dev.openfeature.sdk.EvaluationContext;
+import dev.openfeature.sdk.FeatureProvider;
+import dev.openfeature.sdk.ProviderEvaluation;
+import dev.openfeature.sdk.exceptions.FlagNotFoundError;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.Map;
+import java.util.function.Function;
+
+/**
+ * First match strategy.
+ * Return the first result returned by a provider. Skip providers that indicate they had no value due to
+ * FLAG_NOT_FOUND.
+ * In all other cases, use the value returned by the provider.
+ * If any provider returns an error result other than FLAG_NOT_FOUND, the whole evaluation should error and
+ * “bubble up” the individual provider’s error in the result.
+ * As soon as a value is returned by a provider, the rest of the operation should short-circuit and not call
+ * the rest of the providers.
+ */
+@Slf4j
+public class FirstMatchStrategy extends BaseStrategy {
+
+    public FirstMatchStrategy(Map<String, FeatureProvider> providers) {
```

Comment:
This might be a problem, as the map is generated in the multi-provider; we need these values to be generated before we pass the strategy to the provider. Maybe we should also put those as parameters to the `evaluate` method else it will be hard to use any other strategy than the default

Reply:
I updated it a bit. The providers can be reached via _BaseStrategy_, as done at [MultiProviderTest](https://github.com/liran2000/java-sdk-contrib/blob/daeefe93d86910065bbdcd38f641d0d353c4c506/providers/multiprovider/src/test/java/dev/openfeature/contrib/providers/multiprovider/MultiProviderTest.java#L172). What do you think ?

Reply:
i am not sure, now we do have to separate maps where the state needs to be in sync. this might cause problems. we should centralize this on the provider, what if we change the signature of the `evaluate` method like

```java
public interface Strategy {
    <T> ProviderEvaluation<T> evaluate(Map<...> providers, String key, T defaultValue, EvaluationContext ctx,
           Function<FeatureProvider, ProviderEvaluation<T>> providerFunction);
}

Reply:
it is commonly the same providers instance at the multi-provider and the strategy, and it is built as unmodifiable when initialized. Why can this be a problem ? I am fine with your approach above too, just want to understand better first.

another thing, reg. _ProviderEvaluation<T>> providerFunction_ as parameter:
it is pretty straight-forward to implement with it, but I have some concerns that if a custom strategy implementation want to do change the function call and/or arguments, it will limit it. Other approach can be to pass the type. e.g. Boolean.class instead of the function, and convert it to the specific evaluation method. What do you think ?

Reply:
if i eg. switch to the FirstSuccessful implementation my code would look like 
```java
var providers = List.of(...);
var strategy = new FirstSuccessful(providers);
var provider = new MultiProvider(providers, strategy);
```

but that means that theoretically the providers list could be a different one, or am i missing here something?

Regarding your providerFunction, i do think that this is the clearest way also maintainence wise - trying to hassle with type information adds unnecessary overhead in my opinion

Reply:
ok, updated accordingly.